### PR TITLE
IAM Support: Roles, Role Policies and Instance Profiles

### PR DESCRIPTION
**Pull request by tyler-ball**
I updated https://github.com/chef/chef-provisioning-aws/pull/343 with the following:

* Replaced the discrete `aws_iam_role_policy` resource with `inline_policies` attribute on the Role.  This reflects the AWS IAM console and the language used in the SDK.
* Expanded test coverage
* Rebased off master

I'm creating a new PR so that @randomcamel can review the last commit in this review (my changes from #343)

\cc @Fodoj

### CODE REVIEW

File: spec/aws_support.rb [hunk 1]
```diff
@@ -200,11 +200,11 @@ def self.included(context)
     Chef::Provisioning::AWSDriver::Resources.constants.each do |resource_class|
       resource_class = Chef::Provisioning::AWSDriver::Resources.const_get(resource_class)
       resource_name = resource_class.resource_name
-      define_method("update_an_#{resource_name}") do |name, expected_updates={}|
-        AWSSupport::Matchers::UpdateAnAWSObject.new(self, resource_class, name, expected_updates)
+      define_method("update_an_#{resource_name}") do |name, expected_updates={}, &block|
```
Comment (tyler-ball) [inline spec/aws_support.rb]: This is one of my spec changes - you can now pass a block to the `update`, `create`, and `match` matchers which will be provided the `aws_object`.  This allows for custom matchers on the aws object.  See https://github.com/chef/chef-provisioning-aws/pull/358/files#diff-fcb8f9a6302c2b7fa309abf8b8942159R92

**Comment by Fodoj** [comment]
I still disagree with inline policies decision, because it's not what is used inside SDK and API, but as long as IAM roles finally can be used it's okay ;)

File: spec/aws_support/matchers/create_an_aws_object.rb [hunk 2]
```diff
@@ -8,17 +8,20 @@ class CreateAnAWSObject
       include RSpec::Matchers::Composable
       include AWSSupport::DeepMatcher
 
-      def initialize(example, resource_class, name, expected_values)
+      def initialize(example, resource_class, name, expected_values, custom_matcher)
```
Comment (randomcamel) [inline spec/aws_support/matchers/create_an_aws_object.rb]: Maybe a comment noting that `custom_matcher` is a block? One less thing to track down, for anyone spelunking in the code.

**Comment by randomcamel** [comment]
LGTM, if specs pass, :shipit:

**Comment by tyler-ball** [comment]
I don't know why those tests are having trouble in Travis - I can get them passing locally

**Pull request merged**